src/error_diffusion.jl

```diff
@@ -14,147 +16,164 @@ julia> cs = ColorSchemes.PuOr_7.colors; # using ColorSchemes.jl for color palett
 
 julia> dither(img, alg, cs);
 ```
+
+!!! note "Color images"
+    For color images, two colorspaces are used when applying the algorithm:
+    1) The `Lab` colorspace is used to find the closest color in the color scheme
+        according to the provided color distance metric.
+    2) The colorspace in the optional argument `colorspace=XYZ` is used to propagate the error.
+        Typically a linear colorspace should be used, such as the default value of `XYZ`.
+        Another common choice is `RGB`, which is most commonly used for dithering
+        but strictly speaking is not a linear colorspace.
 """
 
-_error_diffusion_kwargs = """
+# Automate method docstrings
+const _error_diffusion_args = "[colorspace=XYZ]; clamp_error=true"
+const _error_diffusion_details = """
+# Arguments
+- `colorspace`: Color space in which the error is diffused. Defaults to `XYZ`.
+
 # Keyword arguments
-- `color_space`: Color space in which the error is diffused.
-    Only used when dithering with a color palette. Defaults to `XYZ`.
-    To replicate the output of other dithering libraries, set this to `RGB`.
 - `clamp_error::Bool`: Clamp accumulated error on each pixel within limits of colorant
-    type `color_space` before looking up the closest color. Defaults to `true`.
+    of type `colorspace` before looking up the closest color. Defaults to `true`.
 """
 
-struct ErrorDiffusion{F<:AbstractMatrix,C} <: AbstractDither
-    filter::F
+struct ErrorDiffusion{C,I,V} <: AbstractDither
+    inds::I # indices of filter
+    vals::V # values of filter
     clamp_error::Bool
 end
-function ErrorDiffusion(filter; color_space=XYZ, clamp_error=true)
-    return ErrorDiffusion{typeof(filter),float32(color_space)}(filter, clamp_error)
+function ErrorDiffusion(colorspace, inds, vals, clamp_error)
+    length(inds) != length(vals) &&
+        throw(ArgumentError("Lengths of filter indices and values don't match."))
+    return ErrorDiffusion{float32(colorspace),typeof(inds),typeof(vals)}(
+        inds, vals, clamp_error
+    )
+end
+function ErrorDiffusion(
+    filter::AbstractMatrix, offset::Integer, colorspace=XYZ; clamp_error=true
```

Comment:
I'll move colourspaces out of the `ErrorDiffusion` struct and into the call to `dither` in another PR.